src/Rotations.jl

```diff
@@ -0,0 +1,148 @@
+
+project_tangent!(S::Rotations, w, x, v) = w .= (transpose(v).-v)./2
+
+function exp!(S::Rotations, y, x, v)
+    y .= x * exp(v)
+    return y
+end
+
+@doc doc"""
+    log!(M, v, x, y)
+
+compute the logarithmic map on the [`Rotations`](@ref) manifold
+`M`$=\mathrm{SO}(n)$, which is given by
+```math
+\operatorname{log}_{x} y =
+  \frac{1}{2} \bigl(\operatorname{Log}(x^{\mathrm{T}}y)
+  - (\operatorname{Log} x^{\mathrm{T}}y)^{\mathrm{T}}),
+```
+where $\operatorname{Log}$ denotes the matrix logarithm
+and save the result to `v`.
+"""
+function log!(S::Rotations, v::TV, x, y) where TV
+    U = transpose(x) * y
+    if (x != y) && (abs.(U) == one(U))
+        throw( ErrorException("The points $x and $y are antipodal, thus these input parameters are invalid.") )
+    end
+    # MMatrix doesn't have `log` defined
+    U1 = TV(real(log(Array(U))))
+    v .= (U1 .- transpose(U1))./2
+    return v
+end
+
+@doc doc"""
+    retract_polar!(M::Rotations, y, x, v)
+
+This SVD-based retraction is a second-order approximation of the exponential map.
+"""
+function retract_polar!(M::Rotations, y, x, v)
+    A = x + x*v
+    S = svd(A)
+    y .= S.U * transpose(S.V)
+    return y
+end
+
+@doc doc"""
+    retract_qr!(M, y, x, v)
+
+This QR-based retraction is a first-order approximation of the exponential map.
+"""
+function retract_qr!(M::Rotations, y::AbstractArray{T}, x, v) where T
+    A = x + x*v
+    qr_decomp = qr(A)
+    d = diag(qr_decomp.R)
+    D = Diagonal( sign.(d .+ convert(T, 0.5)) )
+    y .= qr_decomp.Q * D
+    return y
+end
+
+retract!(M::Rotations, y, x, v) = M.retract!(M, y, x, v)
+
+zero_tangent_vector(S::Rotations, x) = zero(x)
+zero_tangent_vector!(S::Rotations, v, x) = (v .= zero(x))
+
+"""
+    is_manifold_point(S,x; kwargs...)
+
+checks, whether `x` is a valid point on the [`Rotations`](@ref) `S`,
+i.e. is an array of size [`manifold_dimension`](@ref)`(S)` and represents a
+valid rotation.
+The tolerance for the last test can be set using the ´kwargs...`.
+"""
+function is_manifold_point(S::Rotations{N},x; kwargs...) where {N}
+    if size(x) != (N, N)
+        throw(DomainError(size(x),
+            "The point $(x) does not lie on $S, since its size is not $((N, N))."))
+    end
+    if !isapprox(det(x), 1; kwargs...)
+        throw(DomainError(norm(x), "The determinant of $x has to be +1 but it is $(det(x))"))
+    end
+    if !isapprox(transpose(x)*x, one(x); kwargs...)
+        throw(DomainError(norm(x), "$x has to be orthogonal but it's not"))
+    end
+    return true
+end
+
+"""
+    is_tangent_vector(S,x,v; kwargs... )
+
+checks whether `v` is a tangent vector to `x` on the [`Rotations`](@ref)
+space `S`, i.e. atfer [`is_manifold_point`](@ref)`(S,x)`, `v` has to be of same
+dimension as `x` and orthogonal to `x`.
+The tolerance for the last test can be set using the ´kwargs...`.
+"""
+function is_tangent_vector(S::Rotations{N},x,v; kwargs...) where N
+    is_manifold_point(S,x)
+    if size(v) != (N, N)
```

Comment:
Here `v` is called a vector but is expected to be a matrix.

Reply:
Well, some spaces of matrices are also vector spaces. In particular here, tangent vector is represented by a matrix -- that's how matrix manifolds usually work.

Reply:
I think (as said above) the misconception is, that the _representation_ of a _tangent vector_ can be a matrix (or even a higher dimensional tensor) but its then always meant as an element from the vectorspace (for matrices then the vectorspace R ^{n,n}.

Reply:
Yeah, I'm just thinking that we need to be clear in the documentation that "tangent vector" may have different representations for different manifolds but always refers to an element from a "vector space". In cases like SO(3) where the tangent vectors are sometimes equivalently represented as vectors and sometimes as matrices, we could use dispatch to differentiate between the two.

Reply:
Yes, we should carefully document that and always be exact, for example always really write “tangent vector” e.g. `the tangent vector v is here represented by a matrix` to keep that always clear.